[17.0][IMP] fs_storage: invalidate orm cache when connection fails

Algosse opened the pull request:
Fixes #374 

I adapted a some parts of the code related to [_check_connection](https://github.com/OCA/storage/blob/3c19f00c9b1825afbf5a629e192309a9344e2e04/fs_storage/models/fs_storage.py#L297) from V16.0 (started in #320) and implemented it in V17.0.

Also, I added a check_connection everytime the fs is accessed to detected if it should be generated again or not.

Algosse committed:
[IMP] fs_storage: invalidate orm cache when connection fails

Algosse committed:
[FIX] Do not check connection as it breaks tests

Algosse committed:
[FIX] Line too long

Algosse committed:
[FIX] Typo

lmignon commented:
Thank you @Algosse for all your hard work. I don't work with Odoo version 17. It is therefore difficult for me to give an in-depth review of the proposed changes. I will try to take the time to install a version 17 and understand the problem you are experiencing.

Algosse committed:
[ADD] Use sudo to get fs.storage record otherwise it does not work for non admin users

sbidoul commented:
I have ported all missing commits from 16 to 17 in https://github.com/OCA/storage/pull/397

@Algosse  Could you check if that branch works for you?

lmignon commented:
@Algosse Is your fix still required with the last changes on 16.0?

Algosse commented:
Hello Guys,

Thanks for the forward of V16!

Sorry for the delay, I did not find the time to properly check and run the changes. 

I'm not sure the forward will solve my issue. In my previous tests, I figured the problem was the ormcache of the [get_fs_by_code method](https://github.com/OCA/storage/blob/0f35e5c3d1c94e2c56a7d9a623b7b1e8c12fb9a7/fs_storage/models/fs_storage.py#L217)

If the connection fails, the method keeps returning the same faulty fs. It does not access the fs property so it does not triger the check_connection method.
That is why I updated the get_fs_by_code method so it performs a check everytime (I know it can be costly but it is the best way I found yet). To get the fs, this new method calls the old one, renamed into a private method _get_fs_by_code.


I don't have the time to look at it now, I will be able in a few weeks. 

If someone wants to try, the "best" way I found to crash the connection was to use a fs_storage with SFTP protocol, linked it to a model (for example sale.order) and drag & drop multiple files at once in the chatter (I think Odoo tries to upload files in parallel and fsspec does not like it, at least for SFTP).

john-herholz-dt commented:
@Algosse what is the state of this issue for you? I am facing the same problem. SFTP Connections fail after multiple files.
I have updated latest 17.0 branch but it was not resolved for me.
